main.cpp

```diff
@@ -104,206 +107,452 @@ good to go!
 {
-    return lhs + rhs;
+    *value += rhs;
+
+    return *this;
 }
 
-double DoubleType::subtract( double lhs, double rhs )
+DoubleType& DoubleType::subtract( double rhs )
 {
-    return lhs - rhs;
+    *value -= rhs;
+
+    return *this;
 }
 
-double DoubleType::multiply( double lhs, double rhs )
+DoubleType& DoubleType::multiply( double rhs )
 {
-    return lhs * rhs;
+    *value *= rhs;
+    
+    return *this;
 }
 
-double DoubleType::divide( double lhs, double rhs )
+DoubleType& DoubleType::divide( double rhs )
 {
     if (rhs == 0.0)
     {
-        std::cout << std::endl << "warning, floating point division by zero returns 'inf' !" << std::endl;
+        std::cout << "warning: floating point division by zero!" << std::endl;
     }
 
-    return lhs / rhs;
+    *value /= rhs;
+
+    return *this;
+}
+
+
+DoubleType& DoubleType::add( DoubleType& rhs )
+{
+    return add( *rhs.value );
+}
+
+DoubleType& DoubleType::subtract( DoubleType& rhs )
+{
+    return subtract( *rhs.value );
+}
+
+DoubleType& DoubleType::multiply( DoubleType& rhs )
+{
+    return multiply( *rhs.value );
+}
+
+DoubleType& DoubleType::divide( DoubleType& rhs )
+{
+    return divide( *rhs.value );
+}
+
+
+DoubleType& DoubleType::add( IntType& rhs )
+{
+    return add( *rhs.value );
+}
+
+DoubleType& DoubleType::subtract( IntType& rhs )
+{
+    return subtract( *rhs.value );
+}
+
+DoubleType& DoubleType::multiply( IntType& rhs )
+{
+    return multiply( *rhs.value );
+}
+
+DoubleType& DoubleType::divide( IntType& rhs )
+{
+    return divide( *rhs.value );
+}
+
+
+DoubleType& DoubleType::add( FloatType& rhs )
+{
+    return add( *rhs.value );
+}
+
+DoubleType& DoubleType::subtract( FloatType& rhs )
+{
+    return subtract( *rhs.value );
+}
+
+DoubleType& DoubleType::multiply( FloatType& rhs )
+{
+    return multiply( *rhs.value );
+}
+
+DoubleType& DoubleType::divide( FloatType& rhs )
+{
+    return divide( *rhs.value );
 }
 
 //==============================================================================
-int IntType::add( int lhs, int rhs )
+IntType& IntType::add( int rhs )
 {
-    return lhs + rhs;
+    *value += rhs;
+
+    return *this;
 }
 
-int IntType::subtract( int lhs, int rhs )
+IntType& IntType::subtract( int rhs )
 {
-    return lhs - rhs;
+    *value -= rhs;
+
+    return *this;
 }
 
-int IntType::multiply( int lhs, int rhs )
+IntType& IntType::multiply( int rhs )
 {
-    return lhs * rhs;
+    *value *= rhs;
+
+    return *this;
 }
 
-int IntType::divide( int lhs, int rhs )
+IntType& IntType::divide( int rhs )
 {
     if (rhs == 0)
     {
-        std::cout << "error, integer division by zero will crash the program!";
-        std::cout << std::endl << "returning lhs" << std::endl;
-        return lhs;
+        std::cout << "error: integer division by zero is an error and will crash the program!" << std::endl;
+        
+        return *this;
     }
 
-    return lhs / rhs;
+    *value /= rhs;
+    
+    return *this;
+}
```

Comment:
can you D.R.Y. the `return *this;` statement?  it is written twice.
- [ ] revise